File: packages/ts/src/authz/roles.ts
/**
 * RoleRegistry — composable, role management with optional DB loading.
 */

import { RoleExpandingChecker } from "./checker";
import type { Permission } from "./primitives";

/** Protocol for loading roles from an external source (database, API, etc.). */
export interface RoleLoader {
  loadRoles(): Promise<Map<string, Set<string>>>;
  loadHierarchy(): Promise<Map<string, string[]>>;
}

/** Protocol for caching role data. */
export interface RoleCache {
  get(key: string): Promise<Record<string, unknown> | undefined>;
  set(key: string, value: Record<string, unknown>, ttl: number): Promise<void>;
  invalidate(key: string): Promise<void>;
}

/** In-memory cache with TTL tracking. */
export class MemoryRoleCache implements RoleCache {
  private store = new Map<string, { value: Record<string, unknown>; expiresAt: number }>();

  async get(key: string): Promise<Record<string, unknown> | undefined> {
    const entry = this.store.get(key);
    if (!entry) return undefined;
    if (performance.now() > entry.expiresAt) {
      this.store.delete(key);
      return undefined;
    }
    return entry.value;
  }

  async set(key: string, value: Record<string, unknown>, ttl: number): Promise<void> {
    this.store.set(key, { value, expiresAt: performance.now() + ttl * 1000 });
  }

  async invalidate(key: string): Promise<void> {
    this.store.delete(key);
  }
}

/**
 * Composable registry for role definitions.
 *
 * Supports static roles, merging via `include()`, and DB-loaded roles with caching.
 */
export class RoleRegistry {
  private static CACHE_KEY_ROLES = "role_permissions";
  private static CACHE_KEY_HIERARCHY = "role_hierarchy";

  private staticRoles = new Map<string, Set<string>>();
  private staticHierarchy = new Map<string, string[]>();
  private loadedRoles = new Map<string, Set<string>>();
  private loadedHierarchy = new Map<string, string[]>();
  private loader: RoleLoader | undefined;
  private cache: RoleCache | undefined;
  private cacheTtl = 300;

  /** Register a static role. */
  role(
    name: string,
    permissions: Array<string | Permission>,
    options?: { inherits?: string[] },
  ): void {
    this.staticRoles.set(name, new Set(permissions.map(String)));
    if (options?.inherits) {
      this.staticHierarchy.set(name, [...options.inherits]);
    }
  }

  /** Merge another registry (additive: permissions union, hierarchy merge). */
  include(other: RoleRegistry): void {
    for (const [name, perms] of other.staticRoles) {
      const existing = this.staticRoles.get(name);
      if (existing) {
        for (const p of perms) existing.add(p);
      } else {
        this.staticRoles.set(name, new Set(perms));
      }
    }
    for (const [name, children] of other.staticHierarchy) {
      const existing = this.staticHierarchy.get(name);
      if (existing) {
        const set = new Set(existing);
        for (const c of children) set.add(c);
        this.staticHierarchy.set(name, [...set]);
      } else {
        this.staticHierarchy.set(name, [...children]);
      }
    }
  }

  /** Configure DB loading with optional caching. */
  withLoader(loader: RoleLoader, options?: { cache?: RoleCache; cacheTtl?: number }): void {
    this.loader = loader;
    this.cache = options?.cache;
    this.cacheTtl = options?.cacheTtl ?? 300;
  }

  /** Load roles from the configured loader (via cache if available). */
  async load(): Promise<void> {
    if (!this.loader) return;

    let roles: Map<string, Set<string>> | undefined;
    let hierarchy: Map<string, string[]> | undefined;

    if (this.cache) {
      const cachedRoles = await this.cache.get(RoleRegistry.CACHE_KEY_ROLES);
      const cachedHierarchy = await this.cache.get(RoleRegistry.CACHE_KEY_HIERARCHY);
      if (cachedRoles && cachedHierarchy) {
        roles = new Map(
          Object.entries(cachedRoles).map(([k, v]) => [k, new Set(v as string[])]),
        );
        hierarchy = new Map(
          Object.entries(cachedHierarchy).map(([k, v]) => [k, v as string[]]),
        );
      }
    }

    if (!roles) {
      roles = await this.loader.loadRoles();
      hierarchy = await this.loader.loadHierarchy();
      if (this.cache) {
        const rolesObj: Record<string, string[]> = {};
        for (const [k, v] of roles) rolesObj[k] = [...v];
        const hierObj: Record<string, string[]> = {};
        for (const [k, v] of (hierarchy ?? new Map())) hierObj[k] = v;
        await this.cache.set(RoleRegistry.CACHE_KEY_ROLES, rolesObj, this.cacheTtl);
        await this.cache.set(RoleRegistry.CACHE_KEY_HIERARCHY, hierObj, this.cacheTtl);
      }
    }

    this.loadedRoles = roles;
    this.loadedHierarchy = hierarchy ?? new Map();
  }

  /** Invalidate cache and re-load from the loader. */
  async reload(): Promise<void> {
    if (this.cache) {
      await this.cache.invalidate(RoleRegistry.CACHE_KEY_ROLES);
      await this.cache.invalidate(RoleRegistry.CACHE_KEY_HIERARCHY);
    }
    await this.load();
  }

  private mergedRoles(): Map<string, Set<string>> {
    const merged = new Map(this.loadedRoles);
    for (const [name, perms] of this.staticRoles) {
      merged.set(name, perms); // static wins
    }
    return merged;
  }

  private mergedHierarchy(): Map<string, string[]> {
    const merged = new Map(this.loadedHierarchy);
    for (const [name, children] of this.staticHierarchy) {
      merged.set(name, children); // static wins
    }
    return merged;
  }

  /** Produce a configured RoleExpandingChecker from current state. */
  buildChecker(): RoleExpandingChecker {
    return new RoleExpandingChecker({
      rolePermissions: this.mergedRoles(),
      hierarchy: this.mergedHierarchy(),
    });
  }
}

File: packages/rust/src/authz/roles.rs
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use std::time::{Duration, Instant};

use async_trait::async_trait;
use tokio::sync::RwLock;

use crate::errors::AuthError;
use super::checker::RoleExpandingChecker;

// ---------------------------------------------------------------------------
// Traits
// ---------------------------------------------------------------------------

/// Loads role definitions and hierarchy from an external source (database,
/// config service, etc.).
#[async_trait]
pub trait RoleLoader: Send + Sync {
    /// Load role name -> permission strings mapping.
    async fn load_roles(&self) -> Result<HashMap<String, HashSet<String>>, AuthError>;
    /// Load role name -> inherited role names mapping.
    async fn load_hierarchy(&self) -> Result<HashMap<String, Vec<String>>, AuthError>;
}

/// Simple cache interface used by [`RoleRegistry`] to avoid repeated loads.
#[async_trait]
pub trait RoleCache: Send + Sync {
    async fn get(&self, key: &str) -> Option<String>;
    async fn set(&self, key: &str, value: &str, ttl: u64) -> Result<(), AuthError>;
    async fn invalidate(&self, key: &str) -> Result<(), AuthError>;
}

// ---------------------------------------------------------------------------
// MemoryRoleCache
// ---------------------------------------------------------------------------

struct CacheEntry {
    value: String,
    expires_at: Instant,
}

/// In-memory TTL cache suitable for single-process deployments.
pub struct MemoryRoleCache {
    store: RwLock<HashMap<String, CacheEntry>>,
}

impl MemoryRoleCache {
    pub fn new() -> Self {
        Self {
            store: RwLock::new(HashMap::new()),
        }
    }
}

impl Default for MemoryRoleCache {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl RoleCache for MemoryRoleCache {
    async fn get(&self, key: &str) -> Option<String> {
        let store = self.store.read().await;
        store.get(key).and_then(|entry| {
            if Instant::now() < entry.expires_at {
                Some(entry.value.clone())
            } else {
                None
            }
        })
    }

    async fn set(&self, key: &str, value: &str, ttl: u64) -> Result<(), AuthError> {
        let mut store = self.store.write().await;
        store.insert(
            key.to_string(),
            CacheEntry {
                value: value.to_string(),
                expires_at: Instant::now() + Duration::from_secs(ttl),
            },
        );
        Ok(())
    }

    async fn invalidate(&self, key: &str) -> Result<(), AuthError> {
        let mut store = self.store.write().await;
        store.remove(key);
        Ok(())
    }
}

// ---------------------------------------------------------------------------
// RoleRegistry
// ---------------------------------------------------------------------------

/// Central registry that combines static role definitions with optional
/// dynamic loading and caching.
pub struct RoleRegistry {
    /// role name -> permission strings
    pub roles: HashMap<String, HashSet<String>>,
    /// role name -> inherited roles
    pub hierarchy: HashMap<String, Vec<String>>,
    loader: Option<Arc<dyn RoleLoader>>,
    cache: Option<Arc<dyn RoleCache>>,
    cache_ttl: u64,
}

impl RoleRegistry {
    pub fn new() -> Self {
        Self {
            roles: HashMap::new(),
            hierarchy: HashMap::new(),
            loader: None,
            cache: None,
            cache_ttl: 300,
        }
    }

    /// Register a role with its granted permissions and optional parent roles
    /// it inherits from.
    pub fn role(
        &mut self,
        name: &str,
        permissions: &[&str],
        inherits: Option<&[&str]>,
    ) -> &mut Self {
        let perm_set: HashSet<String> = permissions.iter().map(|s| s.to_string()).collect();
        self.roles
            .entry(name.to_string())
            .or_default()
            .extend(perm_set);

        if let Some(parents) = inherits {
            let parent_list: Vec<String> = parents.iter().map(|s| s.to_string()).collect();
            self.hierarchy
                .entry(name.to_string())
                .or_default()
                .extend(parent_list);
        }

        self
    }

    /// Merge all roles and hierarchy entries from another registry into this one.
    pub fn include(&mut self, other: &RoleRegistry) -> &mut Self {
        for (role_name, perms) in &other.roles {
            self.roles
                .entry(role_name.clone())
                .or_default()
                .extend(perms.clone());
        }
        for (role_name, parents) in &other.hierarchy {
            self.hierarchy
                .entry(role_name.clone())
                .or_default()
                .extend(parents.clone());
        }
        self
    }

    /// Attach a dynamic loader and optional cache.
    pub fn with_loader(
        &mut self,
        loader: Arc<dyn RoleLoader>,
        cache: Option<Arc<dyn RoleCache>>,
        cache_ttl: u64,
    ) -> &mut Self {
        self.loader = Some(loader);
        self.cache = cache;
        self.cache_ttl = cache_ttl;
        self
    }

    /// Load roles from the configured loader and merge them with the
    /// statically registered roles.
    pub async fn load(&mut self) -> Result<(), AuthError> {
        let loader = match &self.loader {
            Some(l) => Arc::clone(l),
            None => return Ok(()),
        };

        let loaded_roles = loader.load_roles().await?;
        let loaded_hierarchy = loader.load_hierarchy().await?;

        for (role_name, perms) in loaded_roles {
            self.roles.entry(role_name).or_default().extend(perms);
        }
        for (role_name, parents) in loaded_hierarchy {
            self.hierarchy.entry(role_name).or_default().extend(parents);
        }

        Ok(())
    }

    /// Build a [`RoleExpandingChecker`] from the current state of the registry.
    pub fn build_checker(&self) -> RoleExpandingChecker {
        RoleExpandingChecker::new(self.roles.clone(), self.hierarchy.clone())
    }
}

impl Default for RoleRegistry {
    fn default() -> Self {
        Self::new()
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn register_roles_and_build_checker() {
        let mut registry = RoleRegistry::new();
        registry
            .role("viewer", &["docs:read"], None)
            .role("editor", &["docs:write"], Some(&["viewer"]))
            .role("admin", &["users:delete"], Some(&["editor"]));

        let checker = registry.build_checker();

        // admin -> editor -> viewer
        let effective = checker.effective_roles(&["admin".to_string()]);
        assert!(effective.contains("admin"));
        assert!(effective.contains("editor"));
        assert!(effective.contains("viewer"));
    }

    #[test]
    fn include_merges_registries() {
        let mut base = RoleRegistry::new();
        base.role("viewer", &["docs:read"], None);

        let mut extra = RoleRegistry::new();
        extra.role("editor", &["docs:write"], Some(&["viewer"]));

        base.include(&extra);

        assert!(base.roles.contains_key("editor"));
        assert!(base.hierarchy.contains_key("editor"));
    }

    #[tokio::test]
    async fn memory_cache_basic() {
        let cache = MemoryRoleCache::new();
        cache.set("key", "value", 60).await.unwrap();
        assert_eq!(cache.get("key").await, Some("value".to_string()));

        cache.invalidate("key").await.unwrap();
        assert_eq!(cache.get("key").await, None);
    }

    #[tokio::test]
    async fn memory_cache_ttl_expiry() {
        let cache = MemoryRoleCache::new();
        // TTL of 0 means it expires immediately.
        cache.set("key", "value", 0).await.unwrap();
        // Instant comparison: expires_at = now + 0s, so now >= expires_at.
        // The get implementation uses strict less-than, so this should be None.
        tokio::time::sleep(std::time::Duration::from_millis(5)).await;
        assert_eq!(cache.get("key").await, None);
    }
}
